drivers/storage/nvme/NvmeProbe.c

```diff
@@ -0,0 +1,645 @@
+    if (!NVMeSetupPCI(pciNvmeDeviceInfo->Bus, pciNvmeDeviceInfo->Slot, &pciData)){
+        DbgPrint("Failed to setup the NVMe PCIe device\n");
+        return FALSE;
+    }
+    
+    /* BAR 0 size */
+    barSize = GetBarSize(pciNvmeDeviceInfo);
+    if (barSize == 0){
+        DbgPrint("Failed to get barsize of NVMe Device\n");
+        return FALSE;
+    }
+    pAE->CtrlRegSize = barSize;
+
+    /* TODO: We should ideally check with what the scsiport has offered as ranges */
+
+    /* MMap BAR0 */
+    PNVMe_CONTROLLER_REGISTERS pNvmeCtrlRegs = (PNVMe_CONTROLLER_REGISTERS)MmMapIoSpace(barAddress, barSize, FALSE);
+    pAE->pCtrlRegister = pNvmeCtrlRegs;
+    DbgPrint("%s BAR0: %p BarSize: 0x%lx pNvmeCtrlRegs: %p\n", __FUNCTION__, (PVOID)barAddress.QuadPart, barSize, pNvmeCtrlRegs);
+
+    /* Read Capabality register */
+    CAP.HighPart = READ_REGISTER_ULONG((PULONG)(&pAE->pCtrlRegister->CAP.HighPart));
+    CAP.LowPart = READ_REGISTER_ULONG((PULONG)(&pAE->pCtrlRegister->CAP.LowPart));
+
+    /* Store whatever is needed for later */
+    pAE->originalVersion.value = READ_REGISTER_ULONG((PULONG)&pAE->pCtrlRegister->VS);
+    pAE->uSecCrtlTimeout       = (ULONG)(CAP.TO * MIN_WAIT_TIMEOUT);
+    pAE->uSecCrtlTimeout       = (pAE->uSecCrtlTimeout == 0) ? MIN_WAIT_TIMEOUT : pAE->uSecCrtlTimeout;
+    pAE->uSecCrtlTimeout      *= MILLI_TO_MICRO;
+    pAE->strideSz              = CAP.DSTRD;
+    DbgPrint("NVMeFindAdapter: Stride Size set to 0x%x\n", pAE->strideSz);
+
+    /* Mask interrutps */
+    WRITE_REGISTER_ULONG(&pNvmeCtrlRegs->IVMS, -1);
+
+    /* Lets alloc and Init Admin queues */
+    if (!NvmeAllocAndInitAdminQueues(pAE)){
+        DbgPrint("Failed to Allocate and Init Admin queues\n");
+        return FALSE;
+    }
+    
+    return TRUE;
+}
+
+/* TODO: Add functionality for passing CNS */
+BOOLEAN
+NVMeSendIdentifyController(PNVME_DEVICE_EXTENSION pDevExtension)
+{
+    if (pDevExtension == NULL)
+        return FALSE;
+
+    NVMe_COMMAND nvmeCmd = {0};
+    NVMe_COMPLETION_QUEUE_ENTRY cmpEntry = {0};
+
+    nvmeCmd.CDW0.OPC = ADMIN_IDENTIFY;
+    nvmeCmd.PRP1     = pDevExtension->QueueInfo.SubQueueInfo->PRPListStartPhysical.QuadPart;
+    ((PADMIN_IDENTIFY_COMMAND_DW10)&nvmeCmd.CDW10)->CNS = CNS_IDENTIFY_CNTLR;
+
+    if (!NVMeIssueCmd(pDevExtension, &nvmeCmd, &cmpEntry, FALSE)) {
+        DbgPrint("Failed to issue Identify controller\n");
+        return FALSE;
+    }
+
+    RtlCopyMemory(&pDevExtension->controllerIdentifyData, 
+                  pDevExtension->QueueInfo.SubQueueInfo->PRPListStartVirtual, 
+                  sizeof(ADMIN_IDENTIFY_CONTROLLER));
+
+    DbgPrint("Sucessfully identified NVMe controller 0x%x 0x%x\n", pDevExtension->controllerIdentifyData.VID, pDevExtension->controllerIdentifyData.SSVID);
+
+    /* TODO: Currenly we are limiting IOs to 1MB but if the controller somehow demands lower max transfer we need to set that up here */
+    return TRUE;
+}
+
+BOOLEAN
+CreateNS(PNVME_DEVICE_EXTENSION pDevExt)
```

Comment:
Is this called from anywhere?

Reply:
Yes this should be called during Init if there's no namespace on the target. Currently its not called from anywhere